Predef maintenance

paulp opened the pull request:
This supersedes #1986. Predef maintenance. Deprecates things which merit deprecation; applies implicit value classes more consistently. Review by @retronym.

paulp commented on src/library/scala/Predef.scala:
```diff
@@ -422,3 +411,31 @@ object Predef extends LowPriorityImplicits {
+  @deprecated("Use the method in scala.Console", "2.11.0") def readf1(format: String)             = Console.readf1(format)
+  @deprecated("Use the method in scala.Console", "2.11.0") def readf2(format: String)             = Console.readf2(format)
+  @deprecated("Use the method in scala.Console", "2.11.0") def readf3(format: String)             = Console.readf3(format)
+}
```
Okay, I guess I have to retarget this group to the undeprecated one.

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1666/

paulp commented:
I deleted most of the jenkins noise because I'm sick of looking at it. "Success" is all that remains.

soc commented:
Could you comment on the `readX` deprecations in Console? I don't understand the need for these changes.

Additionally, is the `package object xsbt` stuff temporary, until SBT is fixed (I really hope so)? Wouldn't it make more sense to fix SBT first, and then remove the methods without the need for such hacks?

retronym commented:
SBT is fixed, but we need to get 0.12.3 released and used before it helps.

soc commented:
... which makes me wonder why this trivial issue has not been fixed on the SBT side since May 2011.

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1672/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1674/

paulp commented:
The deprecations in Console are because I don't think any of those methods belong anywhere near the standard library - look at the range of questionable assumptions embedded in them; hardcoded strings, return values like (Any, Any, Any), no useful error checking or means to handle malformed input - but I did not know what alternative to propose in the deprecation message, so I pushed them into outward to a location isolated from the rest of it, which will make them easier to get rid of later.

Or I'd be happy to say "this method will be removed" on all of them if I can get away with it.

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2406/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2408/

soc commented:
Weird. Somehow my comments disappeared...

> I won't ask why scaladoc needs that sort of helping hand...

Well, if scaladoc had to check every single word in a deprecation message whether it was some valid type, method, value, constant, ... scaladoc would be even slower.

> That is better. I do wish I could give a more concrete reference than the implied "now go google string interpolation".

You could probably link to the article in docs.scala-lang.org. By the way, the deprecation annotation should be added to both the old method and the new implicit class, so that there is a chance that we can get rid of it in the future.

paulp commented:
Anything doing anything makes it slower - there is no information in that statement. Scaladoc is slow because of the quality of the code, and it doing something useful and computationally trivial like identifying known classes inside fixed strings does not deserve to be collateral damage of its general slowness.

JamesIry commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 13255457)
:cat: Roger! Rebuilding pr-rangepos-per-commit, pr-checkin-per-commit for 3ef366e0, 952fe9cd, 75aa5ab9, aca412d3, efb30d40, cac0bd59, 04f151d4. :rotating_light:

JamesIry commented:
The kitteh is happy, @retronym are you happy?

retronym commented on src/library/scala/Predef.scala:
```diff
@@ -260,17 +245,58 @@ object Predef extends LowPriorityImplicits {
+     *  (@see java.lang.String.format).
+     */
+    @inline def formatted(fmtstr: String): String = fmtstr format self
+  }
```
Implicit classes with `val` parameters (which means all the `AnyVal` derived ones) should have obfuscated parameter names. Otherwise we're introducing an implicit from `Any => { def self: ? }`. See `__resultOfEnsuring` above.

paulp replied:
Good point.

scala-jenkins commented:
SPURIOUS ABORT? -- PLS REBUILD/pr-rangepos-per-commit@efb30d40da874c43a07783ceae988bcae0176355

scala-jenkins commented:
(kitty-note-to-self: ignore 13262917)
:cat: Roger! Rebuilding pr-rangepos-per-commit for efb30d40. :rotating_light:

paulp commented:
Updated: likely controversial part involving CharSequence joining the ranks of the Any-parented removed; all the implicit class fields given low-collision-rate names. @retronym or @JamesIry ?

retronym commented on src/library/scala/io/AnsiColors.scala:
```diff
@@ -0,0 +1,53 @@
+package scala
+package io
+
+trait AnsiColors {
```
I think the singular works better, e.g. `AnsiColor.RED`.

paulp committed:
Maintenance of Predef.

1) Deprecates much of Predef and scala.Console, especially:

 - the read* methods (see below)
 - the set{Out,Err,In} methods (see SI-4793)

2) Removed long-deprecated:

 - Predef#exit
 - Predef#error should have gone, but could not due to sbt

At least the whole source base has now been future-proofed
against the eventual removal of Predef#error.

The low justification for the read* methods should be readily
apparent: they are little used and have no call to be in global
namespace, especially given their weird ad hoc semantics and
unreasonably tempting names such as readBoolean().

3) Segregated the deprecated elements in Predef from the part
which still thrives.

4) Converted all the standard Predef implicits into implicit
classes, value classes where possible:

  - ArrowAssoc, Ensuring, StringFormat, StringAdd, RichException (value)
  - SeqCharSequence, ArrayCharSequence (non-value)

Non-implicit deprecated stubs prop up the names of the
formerly converting methods.

paulp commented:
@retronym I renamed it to AnsiColor. How do you likes it.

som-snytt commented:
Would this be the time to ask, whatever happened to Scala constants as White instead of WHITE.  I'm only asking as a partest -colors client.

paulp commented:
It's a good time to ask in a "what a great followup commit that might make, but my does this one look good to me exactly as offered" sort of way.

retronym commented:
The only reservation I have is in adding a non-deprecated trailt `DeprecatedPredef`. The parentage is part of the API, we don't have a means to express that we only want to expose its members, but not its nominal type. It seems nice to coral the methods like this, but it's also a bit redundant as they are annotated with `@deprecated`.

WDYT?

paulp commented:
It's something of a shame to let the implementation call the shots in this regard; I find it noticeably beneficial to segregate methods which exist only for the benefit of the past, not the present. I agree it's suboptimal for the segregationist trait to appear in the API - but can you clarify where that happens? Are we just talking about scaladoc or something else? I would be happy to enhance scaladoc to be a better ignorer.

There are more places than Predef I'd like to perform this segregation, that's why I'd like to move that way. But I can remove it for this PR in the interests of completion.

retronym commented:
Well it's not likely to happen in practice, but `Predef : DeprecatedPredef` would not be source compatible when you removed that trait. Given how strictly we're interpreting binary compatibility these days, we might end up stricter on source-compatibility soon...

paulp commented:
DeprecatedPredef is private[scala]; you can't reference it directly.

paulp commented:
> @deprecated("Use string interpolation instead.", "2.12.0")?

Shoot, I forgot. I'll fix it.

retronym commented:
Righto, I think my paranoia was unjustified. lgtm

soc commented:
> Shoot, I forgot. I'll fix it.

Mhh, is it intended that the old, non-implicit one got the annotation, but not the new one?

paulp commented:
Yes, much as I might like to deprecate any2stringadd, that's not what was happening. Only the pre-implicit-classes mechanism was being deprecated.
